src/server_manager/web_app/ui_components/outline-per-key-data-limit-dialog.ts

```diff
@@ -0,0 +1,341 @@
+                slot="dropdown-content"
+                attr-for-selected="name"
+                selected="${this._initialUnit()}"
+              >
+                <paper-item name="GB">GB</paper-item>
+                <paper-item name="MB">MB</paper-item>
+              </paper-listbox>
+            </paper-dropdown-menu>
+          </div>
+        </div>
+        <div id="buttonsSection">
+          <paper-button id="save" ?disabled=${!this.enableSave} @tap=${this._sendSaveEvent}>${
+        this.localize('save')}</paper-button>
+          <paper-button @tap=${this.close}>${this.localize('cancel')}</paper-button>
+        </div>
+      </paper-dialog>
+    `;
+  }
+
+  private _queryAs<T extends HTMLElement>(selector: string): T {
+    return this.shadowRoot.querySelector(selector) as T;
+  }
+
+  private get _input(): PaperInputElement {
+    return this._queryAs<PaperInputElement>('#dataLimitInput');
+  }
+
+  private _dataLimitValue() {
+    return Number(this._input.value);
+  }
+
+  private _dataLimitType() {
+    return this._queryAs<PaperDropdownMenuElement>('#unitsDropdown').selectedItemLabel as 'GB' |
+        'MB';
+  }
+
+  private _initialUnit() {
+    return this._activeDataLimit()?.unit || 'GB';
+  }
+
+  private _initialValue() {
+    return this._activeDataLimit()?.value.toString() || '';
+  }
+
+  private _activeDataLimit(): DisplayDataAmount|undefined {
+    // Returns the limit which currently is enforced on this key, or undefined if there is none.
+    return this.key?.dataLimit || this.serverDefaultLimit;
+  }
+
+  private async _setCustomLimitTapped() {
+    this.showMenu = !this.showMenu;
+    if (this.showMenu) {
+      await this.updateComplete;
+      this._input.focus();
+    }
+  }
+
+  private _setSaveButtonDisabledState() {
+    this.enableSave = !this._input.invalid;
+  }
+
+  private _sendSaveEvent() {
+    const change = this._dataLimitChange();
+    const eventName = `${change === Change.REMOVED ? 'Remove' : 'Save'}PerKeyDataLimitRequested`;
+
+    this.dispatchEvent(new CustomEvent(eventName, {
+      detail: {ui: this},
+      // Required for the event to bubble past a shadow DOM boundary
+      bubbles: true,
+      composed: true,
+    }));
+  }
+
+  /**
+   * Calculates what type of change, or none at all, the user made.
+   */
+  private _dataLimitChange(): Change {
+    const keyLimit = this.key?.dataLimit;
+    if (this.showMenu) {
+      if (!keyLimit) {
+        return Change.SET;
+      }
+      const inputLimit = this.inputDataLimit();
+      if (inputLimit.value !== keyLimit.value || inputLimit.unit !== keyLimit.unit) {
+        return Change.SET;
+      }
+      return Change.UNCHANGED;
+    }
+    // If we unchecked, then the key will have an active data limit.
+    if (keyLimit) {
+      return Change.REMOVED;
+    }
+    return Change.UNCHANGED;
+  }
+
+  /**
+   * Returns true if the data limit was changed by the user.
+   */
+  public dataLimitChanged() {
```

Comment:
Remove this method and pass the change on the event instead.
No need to pass the dialog on the event.

Reply:
I passed the dialog on the event so that `app` can properly control the dialog.  Otherwise we need to send more events to `appRoot` in order for it to close or reset the dialog.  I find it more readable to have all of that logic in one place in `removePerKeyDataLimit` and `savePerKeyDataLimit`.  It's also nice because then it allows the dialog to define its public API with the normal language access controls.